### PR TITLE
Tunnel: Extend port mapping lookup also for querystring (take 2)

### LINKED ISSUE #203869: Automatic port forward for redirect uri (OAuth 2.0 authorization code flow)
There are command line tools, like Azure `az login`, that uses [OAuth 2.0 authorization code flow](https://learn.microsoft.com/en-us/entra/identity-platform/v2-oauth2-auth-code-flow). This flow involves opening a remote login page and then redirecting back to a local address.

When working remotely, the browsers opens ok on the local computer, but fails when redirecting back, as the port of the command line tool is not locally accessible.

Example of URL that is opened on the browser: `https://login.microsoftonline.com/organizations/oauth2/v2.0/authorize?client_id=x&response_type=code&redirect_uri=http%3A%2F%2Flocalhost%3A44549&scope=https%3A%2F%2Fmanagement.core.windows.net%2F%2F.default+offline_access+openid+profile&state=x&code_challenge=x&code_challenge_method=S256&nonce=x&client_info=1&prompt=select_account`

### Suggestion

I suggest that when opening a local browser, if the URL query parameters contain a local address with port, port forwarding will be created for that port.

In the above example, port 44549 of `redirect_uri` would be automatically forwarded.

Note that the URL in the query parameter is encoded, but a simple regular expression could be used to find the port without real unescaping.

### Workarounds

I'm aware of two workarounds for auto port forwarding for `az login`:

1. Running `az login --debug`.

   This will output the local URL to stdout, which vscode parses and forwards the port (when `remote.autoForwardPortsSource` has the default `output`).

2. Set `remote.autoForwardPortsSource` to `process`.

   This will capture the opening of the local port and forward it.

My goal is that tools like `az login` will work out-of-the-box.

Thanks!

### PR DESCRIPTION
**Pull request by orgads**
When running az login, the URL has localhost in redirect_uri query param. This should trigger automatic port mapping.

Improve localhost port mapping to cover this case as well.

This is a revised and tested version of #203908 which was reverted in #205370.

Fixes #203869.

### CODE REVIEW

**Comment by orgads** [comment]
@alexr00 Sorry, but I still can't test it locally. Remote SSH doesn't work (installed both Remote Development and Remote SSH from vsix files), and TestResolver is not really remote, and it doesn't activate port forwarding.

**Comment by orgads** [comment]
@alexr00 Please review this.

**Review by alexr00 — COMMENTED** [review]
@orgads, I've reverted the original change with https://github.com/microsoft/vscode/pull/205370. Can you make a new PR that include the full feature?

> @alexr00 Sorry, but I still can't test it locally. Remote SSH doesn't work (installed both Remote Development and Remote SSH from vsix files), and TestResolver is not really remote, and it doesn't activate port forwarding.

The test resolver does actually do port forwarding. It forwards to the original port number + 1.

**Comment by orgads** [comment]
@alexr00 Thanks. TestResolver doesn't work out of the box with az login (it only passes the first query param to the host), but I was able to test the feature by ctrl+clicking the full link. Need a bit more tweaking.

**Comment by orgads** [comment]
@alexr00 this time I tested it as much as I could, and also added unit tests (notice the separate commits, if you prefer, I can push the refactoring and tests in a separate PR before merging this one).

One issue I noticed with az login is that Microsoft login verifies that redirect_uri is unchanged, so if the mapped port is not the same, it fails. It looks like the current implementations do keep the port (at least for SSH), but I'm not sure how we should handle cases that don't preserve it. Suggestions?

**Comment by yoshigev** [comment]
> One issue I noticed with az login is that Microsoft login verifies that redirect_uri is unchanged, so if the mapped port is not the same, it fails. It looks like the current implementations do keep the port (at least for SSH), but I'm not sure how we should handle cases that don't preserve it. Suggestions?

Note that I was in contact with the maintainers of az cli on https://github.com/Azure/azure-cli/issues/26556, and they seem to be willing to help solve such issues also at their side.

But as the creation of a the temporary port is done only after the command to open the browser is called, it's probably too late for doing anything. The only way I can think of is supplying some cli command to allocate a temporary port/url and having the code of the cli tool use it.

@alexr00 , is this a common scenario of getting a different port? Is it worth working on? In any case, I think that solving the original bug is a great advance, and the unhandled scenario should not prevent merging the fix.

**Comment by orgads** [comment]
@alexr00 ?

**Comment by alexr00** [comment]
We are in the middle of our [endgame ](https://github.com/microsoft/vscode/issues/205574)release prep right now! I will take a look at this PR next week.

File: src/vs/workbench/electron-sandbox/window.ts
Comment (alexr00) [inline src/vs/workbench/electron-sandbox/window.ts:832]: Why does the query need to be replaced?

Comment (alexr00) [reply src/vs/workbench/electron-sandbox/window.ts:832]: I think I see why it needs to be replaced: If the forward ends up available locally at a different address then the query needs to be updated to point to that new local address.

**Review by alexr00 — COMMENTED** [review]
> @alexr00 , is this a common scenario of getting a different port? Is it worth working on? In any case, I think that solving the original bug is a great advance, and the unhandled scenario should not prevent merging the fix.

The port being different happens when something else is already listening on that port locally. It probably isn't too common. In the case that the port changes, do you think it might be worth not updating the query and simply un-forwarding the port?

File: src/vs/workbench/electron-sandbox/window.ts
Comment (orgads) [reply src/vs/workbench/electron-sandbox/window.ts:832]: Right.

**Comment by orgads** [comment]
> > @alexr00 , is this a common scenario of getting a different port? Is it worth working on? In any case, I think that solving the original bug is a great advance, and the unhandled scenario should not prevent merging the fix.
> 
> The port being different happens when something else is already listening on that port locally. It probably isn't too common. In the case that the port changes, do you think it might be worth not updating the query and simply un-forwarding the port?

I'm really not sure if rejecting the request when the redirect URI is changed is a common scenario. If it is, then I guess your suggestion is preferred.

**Comment by orgads** [comment]
@alexr00 Done, and added tests for all the cases I considered.

**Comment by orgads** [comment]
@alexr00?

**Review by alexr00 — APPROVED** [review]
Looks good, thank you!

**Comment by orgads** [comment]
@alexr00 I prefer not to squash. I used separate commits for a reason. Do you support rebase in this project, or do you want me to rebase?

**Review by benibenj — APPROVED** [review]

**Comment by alexr00** [comment]
We support rebase. I've re-enabled auto-merge with a rebase.

**Comment by orgads** [comment]
Thank you.

**Pull request merged**
